alternator/executor.cc

```diff
@@ -763,15 +763,33 @@ enum class update_tags_action { add_tags, delete_tags };
 static void update_tags_map(const rjson::value& tags, std::map<sstring, sstring>& tags_map, update_tags_action action) {
     if (action == update_tags_action::add_tags) {
         for (auto it = tags.Begin(); it != tags.End(); ++it) {
-            const rjson::value& key = (*it)["Key"];
-            const rjson::value& value = (*it)["Value"];
-            auto tag_key = rjson::to_string_view(key);
-            if (tag_key.empty() || tag_key.size() > 128 || !validate_legal_tag_chars(tag_key)) {
-                throw api_error::validation("The Tag Key provided is invalid string");
+            if (!it->IsObject()) {
+                throw api_error::validation("invalid tag object");
             }
-            auto tag_value = rjson::to_string_view(value);
-            if (tag_value.empty() || tag_value.size() > 256 || !validate_legal_tag_chars(tag_value)) {
-                throw api_error::validation("The Tag Value provided is invalid string");
+            const rjson::value* key = rjson::find(*it, "Key");
+            const rjson::value* value = rjson::find(*it, "Value");
+            if (!key || !key->IsString() || !value || !value->IsString()) {
```

Comment:
are you sure that missing value field doesn't fallback to empty string?

Reply:
Do you mean that a missing value field should be allowed, and mean an empty value?
I'm guessing no, a Value is required, but you're right - I didn't actually test this. I'll add a test.

Reply:
My test confirms that DynamoDB doesn't allow a missing Value (it's not the same as an empty value) -
```
An error occurred (ValidationException) when calling the TagResource operation: 1 validation error detected: Value null at 'tags.1.member.value' failed to satisfy constraint: Member must not be null
```
But the same test also shows that Alternator produces a really ugly message in this case:
```
An error occurred (ValidationException) when calling the TagResource operation: JSON error: condition not met: false
```

So in the next version I'll push shortly, I'll include this additional test, and improve the message in this case.

Reply:
They mention this in docs as allowing NULL, so technically in json it'd mean missing ("Value") key, because empty string is different than NULL. But dunno how it's handled in reality.

Reply:
ok, thanks for testing this

Reply:
I forgot, there is another way to represent null - using keyword, as per json spec: `{"Value": null}`

Reply:
I mis-ran the test - the ugly error message happened on Alternator **without** my patch. After my patch it becomes
```
string Key and Value required
```
So good. And I'll include the new test in my next push.

Reply:
> I forgot, there is another way to represent null - using keyword, as per json spec: `{"Value": null}`

I didn't test "null"... I could but I have other things to do ;-) The empty string case is actually useful (I really tried to do it for the initial_tablets parameter, and was surprised when it didn't work). Sending null or 7 or a hash table instead of a string really looks like a user mistake.